mixer/pkg/runtime/routing/builder.go

```diff
@@ -503,3 +525,114 @@ func (b *builder) getBuilderAndMapperDynamic(_ ast.AttributeDescriptorFinder,
 	}
 	return instBuilder, nil, nil
 }
+
+type ruleAttributes struct {
+	parent    ast.AttributeDescriptorFinder
+	overrides map[string]*descriptor.AttributeManifest_AttributeInfo
+}
+
+func (r *ruleAttributes) GetAttribute(name string) *descriptor.AttributeManifest_AttributeInfo {
+	if info, exists := r.overrides[name]; exists {
+		return info
+	}
+	return r.parent.GetAttribute(name)
+}
+
+// buildRuleCompiler constructs an expression compiler over an extended attribute vocabulary
+// with action name prefixed template output attributes added to the global attribute manifests.
+func (b *builder) buildRuleCompiler(parent ast.AttributeDescriptorFinder, rule *config.Rule) *compiled.ExpressionBuilder {
+	// create attribute type descriptor from action templates
+	attributeDescriptor := make(map[string]*descriptor.AttributeManifest_AttributeInfo)
+	for _, action := range rule.ActionsStatic {
+		// TODO(kuat): rationalize multi-instance check actions
+		if len(action.Instances) > 0 {
+			if template := action.Instances[0].Template; template.Variety == tpb.TEMPLATE_VARIETY_CHECK_WITH_OUTPUT {
+				for _, manifest := range template.AttributeManifests {
+					for attrName, attrInfo := range manifest.Attributes {
+						attributeDescriptor[fmt.Sprintf("%s.output.%s", action.Name, attrName)] = attrInfo
```

Comment:
nit: use `strings.Join` or just `action.Name + ".output." + attrName` here. it should be more efficient.

Reply:
done